src/ising.2.cpp

```diff
@@ -0,0 +1,245 @@
+                point[0] = i; point[1] = j; point[2]=k;
+                
+                int sj = neighbors(N, si, point);
+                H = H - ((J*si[i][j][k]*(sj)) + h*si[i][j][k]);
+            }
+            
+            free (point);
+        
+        return H / 4.;
+}
+
+double deltaE (int N, int *** si, double h, double J, int * point){
+    
+    int sj = neighbors(N, si , point);
+    double dE = -2.0 *(J*si[point[0]][point[1]][point[2]]*sj 
+                                + h*si[point[0]][point[1]][point[2]]);
+    
+    return dE;    
+}
+
+
+void MC_move(int N, int *** si, double h, double J, double &M, double &H, double T){
+    
+    double beta = 1.0/T;
+    int * point = new int [3];
+    int i = (int) rnd()%N; int j = (int)rnd()%N; int k = (int) rnd()%N;
+    
+    point[0] = i; point[1] = j; point[2]=k;
+    
+    
+    double dE = deltaE(N, si, h, J, point);
+    double dm=0.0,dh=0.0;
+    double p = T*log(rnd());
+    if (dE > p)
+    {
+        
+        si[i][j][k] *= -1;
+        dm=2.0*si[i][j][k];
+        dh=dE;
+    }
+    
+    M = M+dm;
+    H = H+dh;
+    
+    free (point);
+    return;
+}
+
+void collect_data(double * data, double M, double H){
+    data[0] = data[0]+M;
+    data[1] = data[1]+H;
+    data[2] = data[2]+M*M;
+    data[3] = data[3]+H*H;
+    data[4] = data[4]+fabs(M);
+}
+
+void configuration(int N, int *** si, string filename){
+    
+    ofstream cfile;
+    cfile.open(filename.c_str());
+    
+    for (int i = 0; i < N; i++)
+        for (int j = 0; j < N; j++)
+            for (int k = 0; k < N; k++)
+            {
+                cfile<<i <<"\t" << j << "\t" << k << "\t" << si[i][j][k] << "\n";
+            }
+            
+            cfile.close();
+        return;
+    
+}
+
+int main(){
+   
+    
+  int N = 10;
+  int MCSteps = 300*N*N*N;
+  double h = 1.0;
+  double J = 2.0;
+  
+  int *** si = new int ** [N];
+  initialize(N, si);
+  
+  double T;
+  double M = Magnetization(N, si);
+  double H = Energy(N, si, h, J);
+  
+  // do{
+        cout << "Running h = " << h << "\n";
+        ofstream tfile;
+        std::ostringstream tf;
+        tf << h <<"_mg.txt" ;
+        std::string tmf = tf.str();
+        tfile.open(tmf.c_str());
+        
+        
+        for (T=1.0; T <= 10.; T+=1.){
+            
+            
+            double * data;
+            data = new double [5];
+            for (int i=0; i < 5; i++) data[i] = 0.0;
+            
+            
+            //Equilibration Steps
+            for (int step = 0; step < MCSteps; step++)
+                MC_move(N, si, h, J, M, H, T);
+            
+            //Production
+            for (int step = 0; step < 0.5*MCSteps; step++){
+                
+                MC_move(N, si, h, J, M, H, T);
+                
+                //recording and updating the M and H values 
+                //Frequency can be varied depending on length/size of simulation
+                collect_data(data, M, H);
```

Comment:
I read a little more about how people typically calculate the statistics for a Monte Carlo run. Turns out we need to average over lattice sweeps, not just individual steps. This is the primary reason the susceptibility and specific heat plots would look wrong.

A single sweep is equal to `N * N * N` steps. One sweep counts as 1 unit of simulation time. The units for a time step are Monte Carlo steps per site.